spg/Animal/lib/Animal.pm

```diff
@@ -36,6 +38,61 @@ if you don't export anything, such as for a purely object-oriented module.
 
 =head1 SUBROUTINES/METHODS
 
+=head2 new
+
+=cut
+
+sub new {
+  my ( $class, $name, $colour ) = @_;
+
+  $colour = default_colour() unless defined $colour;
+
+  return bless { name => $name, colour => $colour }, $class;
+
+}
+
+=head2 set_name
+
+=cut
+
+sub set_name {
+  ref $_[0] or croak "Cannot set name of generic thing";

```

Comment:
you could have put `$_[0]` instead of 'thing' for a slightly more verbose error message